nvidia-drivers: refactor & nvidia_x11 -> nvidia-drivers

codyopel opened the pull request:
- Support for version 355+
  - NVIDIA has added experimental support for EGL, bringing wayland support
    closer (does not work with wayland yet)
- Fix i686 support (for whatever thats worth)
  - Cuda UVM is only supported on x86_64 for 346+
- Rename nvidia_x11 -> nvidia-drivers
  - The NVIDIA driver package contains drivers for several things beyond X.Org
    so referring to it as an X11 driver is inaccurate.
  - Updated references to reflect these change
    - While updating these references it appears several packages are not
      linking against nvidia-drivers correcty, in that they use a specific
      version. Better seperation of kernelspace & userspace is required to
      solve conflicts with some packages.
- Use the long-lived driver as the default
  - Currently the short lived driver is used which may contain breaking changes.
- Add short-lived driver
  - Available as "nvidiaUnstable" in services.xserver.videoDrivers
- Remove support for 173.xx & 304.xx
  - These driver branches require versions of the kernel or xorg-server that are
    no longer supported in nixpkgs.
- Fix patchelf --shrink-rpath
- Install everything explicitly to catch build failures due to changes
- Cleanup shared object version symlinking, _.so_
- Cuda fixes
  - Add missing cuda executables and manpages
- GTK+3 support and used by default (GTK+3 is supported for 346+)
  - nvidia-settings now works correctly on high dpi displays
- Add additional seperation of kernelspace and userspace to fix packaging of
  packages building against the nvidia-drivers. This is currently imcomplete,
  but the ground work for doing so is complete.
- Add nvidia default application profiles
- Cleanup buildinputs

cc @edolstra @vcunat 

All branches have been tested, as well as 32-bit lib support on x86_64.

Versions 173.xx & 304.xx are removed because they are ABI incompatible with
xorg-server 1.17, which is the current version in nixpkgs.

vcunat commented on pkgs/os-specific/linux/nvidia-drivers/default.nix:
```diff
@@ -0,0 +1,229 @@
+
+  patches =
+    optionals (versionMajor == "340") [
+      #./nvidia-340.76-kernel-4.0.patch
```
Why have the patch and reference without actually using it?

vcunat commented on pkgs/os-specific/linux/nvidia-drivers/default.nix:
```diff
@@ -0,0 +1,229 @@
+    inherit
+      version
+      versionMajor;
+    nvenc = (
```
Nitpick: I would personally remove parentheses around these two `if`s in `passthru`. They shouldn't be needed and even visually it's expressed by indentation.

vcunat commented on pkgs/os-specific/linux/nvidia-drivers/default.nix:
```diff
@@ -0,0 +1,229 @@
+  , pango ? null
+  , gtk2Support ? false
+    , gtk2 ? null
+  , gtk3Support ? true
```
I remember Eelco didn't like the dependency on gtk3 by default (a few months ago). I think we should start preferring gtk3 at *some* point &ndash; even in general though I don't think we need a mass switch in cases where both versions are well supported.

vcunat commented:
Niptick we can fix during cherry-picking: there should be an empty line after the first one in the commit message.

codyopel committed:
nvidia-drivers: refactor & nvidia_x11 -> nvidia-drivers

+ Support for version 355+
  - NVIDIA has added experimental support for EGL, bringing wayland support
    closer (does not work with wayland yet)
+ Fix i686 support (for whatever thats worth)
  - Cuda UVM is only supported on x86_64 for 346+
+ Rename nvidia_x11 -> nvidia-drivers
  - The NVIDIA driver package contains drivers for several things beyond X.Org
    so referring to it as an X11 driver is inaccurate.
  - Updated references to reflect these change
    * While updating these references it appears several packages are not
      linking against nvidia-drivers correcty, in that they use a specific
      version. Better seperation of kernelspace & userspace is required to
      solve conflicts with some packages.
+ Use the long-lived driver as the default
  - Currently the short lived driver is used which may contain breaking changes.
+ Add short-lived driver
  - Available as "nvidiaUnstable" in services.xserver.videoDrivers
+ Remove support for 173.xx
  - This driver branch requires a version of xorg-server that is no longer
    supported in nixpkgs.
+ Fix patchelf --shrink-rpath
+ Install everything explicitly to catch build failures due to changes
+ Cleanup shared object version symlinking, *.so*
+ Cuda fixes
  - Add missing cuda executables and manpages
+ GTK+3 support and used by default (GTK+3 is supported for 346+)
  - nvidia-settings now works correctly on high dpi displays
+ Add additional seperation of kernelspace and userspace to fix packaging of
  packages building against the nvidia-drivers. This is currently imcomplete,
  but the ground work for doing so is complete.
+ Add nvidia default application profiles
+ Cleanup buildinputs

cc @edolstra @vcunat

codyopel commented:
@vcunat I updated it with 304.xx, I haven't fully tested it though so the kernel module will need patching with newer kernels.

vcunat commented:
Why did you close this?
